File: vital-core/src/main/java/me/vitalframework/annotation/RequiresSpigot.java
package me.vitalframework.annotation;

import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@ConditionalOnClass(name = "org.bukkit.plugin.java.JavaPlugin")
@Target({ElementType.TYPE, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface RequiresSpigot {
}

File: vital-core/src/main/kotlin/me/vitalframework/RequiresSpigot.kt
package me.vitalframework

import org.springframework.context.annotation.Condition
import org.springframework.context.annotation.ConditionContext
import org.springframework.context.annotation.Conditional
import org.springframework.core.type.AnnotatedTypeMetadata
import org.springframework.stereotype.Component
import org.springframework.util.ClassUtils

/**
 * Convenience-condition to mark a class to only be loaded as a bean, when running as a Spigot plugin.
 * If not running as a Spigot plugin, the annotated bean will not be instantiated by Spring.
 *
 * Must be used in combination with [Conditional] and any [Component] stereotype.
 *
 * ```java
 * @Conditional(RequiresSpigot.class)
 * @Component
 * public class MySpigotBean {
 *   // ...
 * }
 * ```
 */
class RequiresSpigot : Condition {
    override fun matches(
        context: ConditionContext,
        metadata: AnnotatedTypeMetadata,
    ): Boolean = ClassUtils.isPresent("org.bukkit.plugin.java.JavaPlugin", context.classLoader)
}
